Create xplat and Windows native symbol packages

dagood opened the pull request:
On Windows, in Debug configuration, creates symbol packages with the `.pdb` from the native build.

On Linux/OSX, in Release configuration, creates symbol packages with symbols that have been stripped from the release library. On Linux the symbol files are `.so.dbg`, on OSX `.dylib.dwarf`.

The symbol stripping code is from https://github.com/dotnet/coreclr/blob/afdce3a592e/CMakeLists.txt#L279-L340 as well as some simplified platform detection logic from earlier in that file. I also looked at the PRs for stripping coreclr symbols (https://github.com/dotnet/coreclr/pull/3872) and creating coreclr symbol packages (https://github.com/dotnet/coreclr/pull/4218) a lot while making these changes.

I've tested on Win10, OSX 10.11, and Ubuntu 14.04.4. I've uploaded the resulting native symbol packages to https://www.myget.org/gallery/dagood-test-native-symbols-corefx.

I had to rig a few things to test this out (I'll be looking into them further):
* On OSX, I had to force my `RuntimeOS` to `osx.10.10` in `build-packages.sh`. It's being used to filter which packages are created, and my `osx.10.11` resulted in no runtime packages being made.
* On all platforms I passed `/p:ConfigurationGroup=Release` (or `Debug`) to `build-packages`. The command should probably have a switch built in.

/cc @ericstj @weshaggard @brianrob

ellismg commented on src/Native/CMakeLists.txt:
```diff
@@ -99,6 +99,114 @@ if (CMAKE_SYSTEM_NAME STREQUAL Linux)
+    endif(CLR_CMAKE_PLATFORM_UNIX)
+endfunction()
+
+function(install_clr targetName)
```
NIT: I believe I understand why this function is named like it is (since these targets initially came from CoreCLR), but maybe we should instead call it `install_corefx` or `install_library_and_symbols` ?

dagood replied:
Yep, that's the reason, but I agree. I like `install_library_and_symbols`. `System.Native` still uses normal `install` for the static lib and using a more descriptive name for the dynamic lib will make the difference more obvious than `install` vs. `install_corefx`. (And if we need to change `install_...` to do more stuff than just libraries and symbols I think we could just refactor at that point.) Renaming.

dagood commented:
Re: TFS, I realized that the xplat changes won't work without TFS-side changes because of the gap between creating and packing the nuspecs. With that in mind, I'm testing to see if this change at least doesn't break anything until TFS changes get in.

(The Windows natives might work fine because they use `.pdb`.)

ellismg commented on src/Native/Windows/clrcompression/CMakeLists.txt:
```diff
@@ -74,3 +74,7 @@ GENERATE_EXPORT_HEADER( clrcompression
 install (TARGETS clrcompression DESTINATION .)
+
+install (FILES ${CMAKE_CURRENT_BINARY_DIR}/Debug/clrcompression.pdb DESTINATION .
+    CONFIGURATIONS Debug
```
What happens for Windows/Release?

dagood replied:
When I do `build native /p:ConfigurationGroup=Release`, I only get a `dll`, `exp`, and `lib` in `bin\obj`. This install doesn't run, so there's only the `dll` and `lib` in `bin`. When I build the packages, I get the normal library package with the `dll`, but since there's no `pdb` NuGetPack skips making the symbol package.

Do you think that we should have a `pdb` for release builds as well? I'm not familiar enough with native builds to know when we want one and when we can get one. (It would be simple to remove `CONFIGURATIONS Debug` and let the pdb flow through if someone can help me get the pdb created.)

ellismg replied:
> Do you think that we should have a pdb for release builds as well? I'm not familiar enough with native builds to know when we want one and when we can get one.

We should always be producing a `.pdb` for any build we do (either Debug or Release).  If we aren't for release, then we need to get our CMake rules for this thing updated.  @ianhays Did we go out of our way to not produce a PDB for System.IO.Compression when building Release?

Not having a PDB will be a ship blocker (today my understanding is the version of this library that we produce is still built out of TFS and doesn't use CMake).

dagood replied:
Thanks, I read up a bit on it and yeah, that being a blocker makes sense. I'll see if I can get the build creating and packaging the release PDB.

ianhays replied:
The clrcompression cmake build is missing lots of linker and compiler args that it should have before it can be shipped. One of those is the DEBUG/Zi flag set to enable pdb generation. 

I'm working today on getting it to parity with the one we build internally. That will include generation of pdbs for both builds. I'll CC you both on the PR :)

ianhays replied:
https://github.com/dotnet/corefx/pull/7840

dagood replied:
Cool, thanks @ianhays! I'll try out those changes on top of this PR to see if the packaging still finds things where it expects them.

ellismg commented:
> You can check out the symbol packages by clicking the "more" links or directly downloading by URL convention. I have not tried to debug with these symbols--I don't know how to set up a reasonable scenario.

On windows, you could consider using `symchk` (which comes with the debugging tools for windows) to ensure that the symbols in the package are private symbols and line up with the binaries we are producing.

@mikem8361 might know if there is a similar tool on OSX/Linux to ensure that the a symbol file matches up to a `.so`

ellismg commented on src/Native/pkg/dir.targets:
```diff
@@ -0,0 +1,31 @@
+    </File>
+
+    <!-- Include placeholder pdb on non-Windows platforms to ensure package is treated like a symbol package. -->
+    <File Include="$(MSBuildThisFileDirectory)\_.pdb" Condition="'$(PackageTargetRuntime)'!='' AND '$(OSGroup)'!='Windows_NT'">
```
Have we talked with MyGet/NuGet to understand if this sort of thing is reasonable to be doing?  I worry a little bit it's going to cause problems for downstream consumers.

dagood replied:
No, we're trying to work out what the way forward here should be in general, too, so we still need to talk to MyGet. I added you to the thread about this. --But we should definitely talk to them before publishing these to our real feed, even just to make sure they're aware of this.

I don't see how these would cause any problems, though--these packages upload straight to a MyGet-specific feature (skipping the dotnet-core feed), so it seems like it would require quite some effort to notice this, let alone get broken by it. Maybe I don't understand the kind of problems you're thinking of.

dagood commented:
Good idea. I tried out `dwarfdump --uuid` on my OSX box and it returned the same values for corresponding files, so that's a start. That doesn't seem like a rigorous check, though.

`symchk` passes the Windows/Debug dll+pdb pair, but for Windows/Release it says `SYMCHK: clrcompression.dll   FAILED  - No CodeView information found.` (Not that I had a pdb to test against.)

ellismg commented:
> but for Windows/Release it says SYMCHK: clrcompression.dll FAILED - No CodeView information found. (Not that I had a pdb to test against.)

We'll need to get a symbol file produced before we start using this `clrcompression.dll` for real  We should get a bug on it.

dagood commented:
I added a linker flag to create Windows/Release `clrcompression.pdb` and started making a symbol package for it. I also found people suggesting the REF and ICF linker optimizations (https://msdn.microsoft.com/en-us/library/bxwfs976.aspx) so I threw those in for now--let me know if they're not appropriate.

mikem8361 commented:
@mikem8361<https://github.com/mikem8361> might know if there is a similar tool on OSX/Linux to ensure that the a symbol file matches up to a .so
I don’t know of any tools yet.  I’m still learning (even after a year or so) about Linux and OS X.

dagood commented:
https://github.com/dotnet/corefx/pull/7840 does the Windows native pdb packaging, so once that goes in I'll remove the Windows changes I have here to get rid of the merge conflict that'll show up. :smile:

ianhays commented:
@dagood done! PDBs will be produced with clrcompression now.

dagood commented:
@ianhays Thanks! Rebased on master.

There have been some changes to the packaging plan, though, so a big chunk of my PR isn't valid anymore. (And I'm going to be making some buildtools changes.) If someone thinks that adding the xplat stripping to the build without packaging would be helpful I could work on this PR some more to enable that, otherwise I'm going to hold off for now until I rework the symbol packages.

dagood commented:
Rebased this on latest and cleaned up. Previously I'd thought we were going a different direction with the symbol packages, but it turns out this PR is the way we're going forward. At some point we won't need the placeholder `_.pdb` and `src/_._` and I'll remove them then. In general, MyGet is on board and there aren't objections or better solutions from anyone else. (Unless someone here can think of one now. :smile:)

mmitche commented:
@dotnet-bot test Innerloop OSX Release Build and Test (rebooted machine by accident)

dagood commented:
@ellismg Now that the symbols plan is more solid, could you look again? @ericstj Maybe you could take a look too? Thanks!

ericstj commented:
Packaging changes look reasonable to me.  Make sure to diff the nuspecs before/after your change to make sure you only see changes you expect.  I commonly use `msbuild packages.builds /p:BuildPackageLibraryReferences=false /m /p` move the bin/packages/debug/specs directory, apply my changes, rerun that command and diff the old and new specs directory.

ellismg commented on src/Native/pkg/dir.props:
```diff
@@ -22,4 +31,13 @@
+
+  <!-- Add path globs specific to native binaries to exclude unnecessary files in packages. -->
+  <ItemGroup>
+    <AdditionalLibPackageExcludes Include="%2A%2A\%2A.dbg" Condition="'$(OSGroupPath)'=='Linux'" />
```
Can we just use $(SymbolFileExtension) here instead (and have the glob be the same?)

dagood replied:
Good idea, and I think I'll do similar for the other globs. (I'll also move the condition to the `ItemGroup` to continue not adding these on Windows.)

ellismg commented:
@dagood, you will need another rebase to pick up the Open Suse 13.2 package authoring stuff, so you can make changes to that pkgproj as well, I think.

dagood commented:
@ellismg Thanks, rebased again and changed those pkgprojs.

@ericstj Good idea. It indeed added pdbs to Windows Debug and Release nuspecs, and only added the xplat native symbols on the Release build (where they would be stripped into the symbol file). Only other change was that it ordered the `.so` files to be after the `.a` files in the packages that had both, due to `NativeFile` vs `File`.

ellismg commented:
Other than my small comments, LGTM.  Thanks for doing this, Davis.

dagood commented:
@dotnet-bot test Innerloop OSX Debug Build and Test
@dotnet-bot test Innerloop OSX Release Build and Test
(Jenkins reboot.)
